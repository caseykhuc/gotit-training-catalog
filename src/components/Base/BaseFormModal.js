import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import BaseModal from './BaseModal';
import BaseForm from './BaseForm';

class BaseFormModal extends React.Component {
  state = this.props.initialState

  onInputChange = (e) => {
    const { name, value } = e.target;
    const { inputValue } = this.state;

    const newValue = { ...inputValue, [name]: value };
    const { validate } = this.props;

    this.setState(validate ? {
      inputValue: newValue,
      inputError: validate(newValue),
    } : {
        inputValue: newValue,
      });
  }

  onFormSubmit = async () => {
    const { onAction, initialState, validate } = this.props;
    const { inputValue, inputError } = this.state;

    console.log(inputValue);
    if (!_.values(inputValue).some((x) => !_.isEmpty(x))) {
      this.setState({ requestError: 'Empty input' });
      return;
    }

    if (validate && !_.isEmpty(validate(inputValue))) {
      return;
    }

    if (inputError) {
      const res = await onAction(inputValue);
      if (!res.success) {
        this.setState((typeof res.message === 'object')
          ? { inputError: res.message }
          : { inputError: {}, requestError: res.message })
      } else this.setState(initialState)
    }
  }

  onKeyDown = (e) => {
    if (e.key === 'Enter') {
      this.onFormSubmit();
    }
  }

  render() {
    const {
      inputValue, inputError, requestError,
    } = this.state;

    const { fields, title } = this.props;
    return (
      <BaseModal title={title} onAccept={() => this.onFormSubmit()}>
        <BaseForm
          onInputChange={(e) => this.onInputChange(e)}
          inputValue={inputValue}
          inputError={inputError}
          requestError={requestError}
          fields={fields}
          onKeyDown={(e) => this.onKeyDown(e)}
        />
      </BaseModal>
    )
  }
}

BaseFormModal.propTypes = {
  fields: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string,
    value: PropTypes.string,
  })),
  initialState: PropTypes.shape({
    inputValue: PropTypes.object,
    inputError: PropTypes.object,
    requestError: PropTypes.string,
  }),
  onAction: PropTypes.func.isRequired,
  title: PropTypes.string,
  validate: PropTypes.func,
}

export default BaseFormModal;
