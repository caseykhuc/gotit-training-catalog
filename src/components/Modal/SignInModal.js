import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { signinUser } from '../../actions/user';
import BaseFormModal from '../Base/BaseFormModal';

export class SignInModal extends Component {
  initialState = {
    inputValue: {
      username: '',
      password: '',
    },
    inputError: {
    },
    requestError: '',
  }

  fields = [{ name: 'username', type: 'text' }, {
    name: 'password', type: 'password',
  }]

  render() {
    const { initialState, fields } = this;
    const { signinUser } = this.props;
    return (
      <BaseFormModal title="SIGN IN" fields={fields} initialState={initialState} onAction={signinUser} />
    )
  }
}

SignInModal.propTypes = {
  signinUser: PropTypes.func.isRequired,
}

export default connect(null, { signinUser })(SignInModal);
