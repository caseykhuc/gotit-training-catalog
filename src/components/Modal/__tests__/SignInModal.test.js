import React from 'react';
import { shallow } from 'enzyme';
import { SignInModal } from '../SignInModal';
import BaseFormModal from '../../Base/BaseFormModal';

describe('components/Modal/SignInModal', () => {
  let props;
  let wrapper;
  let baseFormModal;

  const update = () => {
    wrapper.update();
    baseFormModal = wrapper.find(BaseFormModal);
  };
  const setup = () => {
    wrapper = shallow(<SignInModal {...props} />);
    update();
  };
  beforeEach(() => {
    props = {
      signinUser: jest.fn(),
    }
  });
  it('should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });
  it('should invoke register when BaseFormModal invoked onAccept ', () => {
    setup();
    baseFormModal.props().onAction();
    expect(props.signinUser).toHaveBeenCalled();
  });
})
