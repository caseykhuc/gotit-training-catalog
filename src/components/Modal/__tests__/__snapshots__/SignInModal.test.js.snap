// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Modal/SignInModal should render correctly 1`] = `ShallowWrapper {}`;

exports[`components/SignInModal should render correctly 1`] = `ShallowWrapper {}`;
